File: typescript/exercises.ts
import exp from "node:constants"
import { open } from "node:fs/promises"
type Predicate<T> = (element: T) => boolean
type Mapper<T, U> = (element: T) => U

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<T, U>(
  arr: T[],
  predicate: Predicate<T>,
  mapper: Mapper<T, U>
): U | undefined {
  const foundElement = arr.find(predicate)
  return foundElement ? mapper(foundElement) : undefined
}

export function* powersGenerator(base: bigint): Generator<bigint> {
  let exponent = 0n
  while (true) {
    let result: bigint = 1n
    for (let i = 0n; i < exponent; i++) {
      result *= base
    }
    yield result
    exponent++
  }
}

export async function meaningfulLineCount(filePath: string): Promise<number> {
  let count = 0
  const fileHandle = await open(filePath, "r")
  try {
    const readStream = fileHandle.createReadStream({ encoding: "utf-8" })
    let data = ""

    for await (const chunk of readStream) {
      data += chunk
      const lines = data.split("\n")
      if (data[data.length - 1] !== "\n") {
        data = lines.pop()!
      } else {
        data = ""
      }

      for (const line of lines) {
        const trimmedLine = line.trim()
        if (trimmedLine.length > 0 && !trimmedLine.startsWith("#")) {
          count++
        }
      }
    }

    if (data.length > 0) {
      const trimmedLine = data.trim()
      if (trimmedLine.length > 0 && !trimmedLine.startsWith("#")) {
        count++
      }
    }
  } finally {
    await fileHandle.close()
  }

  return count
}

interface Sphere {
  kind: "Sphere"
  radius: number
}

interface Box {
  kind: "Box"
  width: number
  length: number
  depth: number
}

export type Shape = Sphere | Box

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return 4 * Math.PI * shape.radius * shape.radius
    case "Box":
      const { width, length, depth } = shape
      return 2 * (width * length + length * depth + width * depth)
  }
}

export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return (4 / 3) * Math.PI * Math.pow(shape.radius, 3)
    case "Box":
      const { width, length, depth } = shape
      return width * length * depth
  }
}

export function toString(shape: Shape): string {
  switch (shape.kind) {
    case "Sphere":
      return `Sphere with radius ${shape.radius}`
    case "Box":
      return `Box with width ${shape.width}, length ${shape.length}, and depth ${shape.depth}`
  }
}

export interface BinarySearchTree<T> {
  size(): number
  insert(value: T): BinarySearchTree<T>
  contains(value: T): boolean
  inorder(): Iterable<T>
  toString(): string
}

export class Empty<T> implements BinarySearchTree<T> {
  insert(value: T): BinarySearchTree<T> {
    return new Node<T>(value, new Empty<T>(), new Empty<T>())
  }

  contains(value: T): boolean {
    return false
  }

  size(): number {
    return 0
  }

  inorder(): Iterable<T> {
    return []
  }

  toString(): string {
    return "()"
  }
}

class Node<T> implements BinarySearchTree<T> {
  constructor(
    private value: T,
    private left: BinarySearchTree<T>,
    private right: BinarySearchTree<T>
  ) {}

  insert(newValue: T): BinarySearchTree<T> {
    if (newValue < this.value) {
      return new Node(this.value, this.left.insert(newValue), this.right)
    } else if (newValue > this.value) {
      return new Node(this.value, this.left, this.right.insert(newValue))
    } else {
      return this
    }
  }

  contains(value: T): boolean {
    if (value === this.value) return true
    if (value < this.value) return this.left.contains(value)
    return this.right.contains(value)
  }

  size(): number {
    return 1 + this.left.size() + this.right.size()
  }

  *inorder(): Iterable<T> {
    yield* this.left.inorder()
    yield this.value
    yield* this.right.inorder()
  }

  toString(): string {
    const leftStr = this.left.toString()
    const rightStr = this.right.toString()
    return `(${leftStr}${this.value}${rightStr})`.replace(/\(\)/g, "")
  }
}

export class BinarySearchTreeFactory {
  static createEmpty<T>(): BinarySearchTree<T> {
    return new Empty<T>()
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

let first_then_apply array predicate consumer =
  match List.find_opt predicate array with 
  | None -> None
  | Some x -> consumer x


let powers_generator base =
  let rec generator_from power () =
    Seq.Cons (power, generator_from (power * base))
  in
  generator_from 1;;

let meaningful_line_count (filename: string) : int =
  let count_lines_in_channel channel =
    let rec aux acc =
      try
        let line = input_line channel in
        let trimmed_line = String.trim line in
        if trimmed_line <> "" && not (String.starts_with ~prefix:"# " trimmed_line) then
          aux (acc + 1) 
        else
          aux acc  
      with End_of_file ->
        acc  
    in
    aux 0
  in
  let in_channel = open_in filename in
  Fun.protect
    (fun () -> count_lines_in_channel in_channel) 
    ~finally:(fun () -> close_in in_channel)  


type shape =
  | Sphere of float 
  | Box of float * float * float  

let volume s =
  match s with
  | Sphere r -> (4.0 /. 3.0) *. Float.pi *. (r ** 3.0) 
  | Box (l, w, h) -> l *. w *. h 

let surface_area s =
  match s with
  | Sphere r -> 4.0 *. Float.pi *. (r ** 2.0) 
  | Box (l, w, h) -> 2.0 *. (l *. w +. l *. h +. w *. h)  

let to_string s =
  match s with
  | Sphere r -> Printf.sprintf "Sphere(radius: %.2f)" r
  | Box (l, w, h) -> Printf.sprintf "Box(length: %.2f, width: %.2f, height: %.2f)" l w h


type 'a binary_search_tree =
  | Empty
  | Node of 'a binary_search_tree * 'a * 'a binary_search_tree

let rec insert x tree =
  match tree with
  | Empty -> Node (Empty, x, Empty)
  | Node (left, value, right) ->
      if x < value then
        Node (insert x left, value, right)  
      else if x > value then
        Node (left, value, insert x right) 
      else
        tree  

let rec lookup x tree =
  match tree with
  | Empty -> false
  | Node (left, value, right) ->
      if x < value then
        lookup x left  
      else if x > value then
        lookup x right 
      else
        true  

let contains x tree =
  lookup x tree  

let rec size tree =
  match tree with
  | Empty -> 0
  | Node (left, _, right) -> 1 + size left + size right

let rec inorder tree =
  match tree with
  | Empty -> []
  | Node (left, value, right) -> inorder left @ [value] @ inorder right
